### PR TITLE
group fix - mirror images group changes to thumbnails

### LINKED ISSUE #4791: groups are going crazy
**Describe the bug**

The group collapsing / expanding can go crazy

**To Reproduce**
1. Go to lighttable
2. Select a collection with 3 pictures
3. Select the 3 pictures and group them
4. Activate the grouped view (big G on top-right of lt)
5. Click on the G of the group leader (should be first pict num = group num) , all ok the image are 
ungrouped
6. Click again on the G of the group leader (should be first pict num = group num) , all ok the image are grouped (repeat 5 & 6 all is ok)
7. Click on G if needed to ungroup the 3 images
8. Now, click on G of second image, it changes the group leader and only 2 images are grouped, see that group number is changed in "image information".
9. It seems impossible at this stage to have all 3 images grouped together except by toggling off the grouped view and enabling it back. 

**Expected behavior**

Do nothing when clicking on G of non group leader ?
Or as of today change the group leader, but ensure the collapsing occurs properly ?

**Screenshots**

**Platform (please complete the following information):**
 - Darktable Version: master
 - OS: Debian
 - 
 - 

**Additional context**

### PR DESCRIPTION
**Pull request by phweyland**
Fixes #4791 fro me.

Some doubts.
- is the way to specify the thumbnails table (always) correct ?
- in views/lighttable it seems there is a lot unused code: button_press(), key_release()... I understand that this part is now treated by dtgtk/thumbnail.c. Is that not to be cleaned up ?

### CODE REVIEW

**Comment by AlicVB** [comment]
Great ! I'll have a look asap.
Yes there's lot of unused code in lighttable. But most is still here because of culling/preview. Once I'll have finished that part of the rewrite, lighttable code will be really more lightweight !

File: src/dtgtk/thumbtable.c [hunk 1]
```diff
@@ -426,6 +426,36 @@ static int _thumbs_remove_unneeded(dt_thumbtable_t *table)
   return changed;
 }
 
+// change thumbnails group id
+void dt_thumbtable_group_change_representative(dt_thumbtable_t *table, const int group_id, const int new_group_id)
+{
+  if(table->mode != DT_THUMBTABLE_MODE_FILEMANAGER) return;
```
Comment (AlicVB) [inline src/dtgtk/thumbtable.c:432]: iirc, it's also possible to "play" with groups on zoomable and filmstrip, so not sure we need this check. (same for lines above)

**Comment by phweyland** [comment]
>  t would be to use a more generic function in `dt_thumbtable_t` that would reload the infos of all or one of the images in the list (`dt_thumbnail_update_infos`).

I've made a try with _image_get_infos() but it seems reserved to overlay (`if(thumb->over == DT_THUMBNAIL_OVERLAYS_NONE) return;`). But yes I that should be possible. It may be a bit slower as you say.

I'll look at this.

File: src/dtgtk/thumbtable.c (hunk 1 above)
Comment (phweyland) [reply src/dtgtk/thumbtable.c:432]: On culling mode there is no way to select several images (at least I haven't found it).
On filmstrip, the G icon doesn't appear> I've also tried the shortcut but that does nothing...

Comment (phweyland) [reply src/dtgtk/thumbtable.c:432]: > it's also possible to "play" with groups on zoomable and filmstrip, so not sure we need this check.

I have doubt here, I use dt_ui_thumbtable(darktable.gui->ui) to identify the thumbtable. Is it valid also for zoomable and filmstrip ?

**Comment by phweyland** [comment]
> (`dt_thumbnail_update_infos`)

The parameter is thumb, but in all cases the known parameter is imgid or groupid.
It works well to mirror group changes in grouping.c routines.

To follow your suggestion the only way is to call `dt_thumbnail_update_infos()` for all visible images on any group change. Is that your suggestion ?

File: src/dtgtk/thumbtable.c (hunk 1 above)
Comment (phweyland) [reply src/dtgtk/thumbtable.c]: I've removed this test. It works now for zoomable and filmstrip.

**Comment by phweyland** [comment]
> Ideally that would even be a new signal like DT_IMAGE_INFOS_CHANGE but that's way more work...

Following this idea, instead of updating thumb's group, we could rise this signal.
The response to that signal would be to update the thumb data. 

The signal DT_SIGNAL_MOUSE_OVER_IMAGE_CHANGE is often used for that purpose (rating, metadata,... changes). Should be replaced by the new signal.

Do I understand your idea properly ?

**Comment by AlicVB** [comment]
> Following this idea, instead of updating thumb's group, we could rise this signal.
> The response to that signal would be to update the thumb data.

Exactly

> The signal DT_SIGNAL_MOUSE_OVER_IMAGE_CHANGE is often used for that purpose (rating, metadata,... changes). Should be replaced by the new signal.

Yes, it would be better, but that will need some testing to be sure not to forget something.

> Do I understand your idea properly ?

yes. I'm a little ashamed to let you fix that, but my dt time is currently fulfilled with the culling rewrite...  Anyway, don't hesitate to ask if you need help !

**Comment by phweyland** [comment]
> I'm a little ashamed to let you fix that, but my dt time is currently fulfilled with the culling rewrite

You have made a huge work here. It's a pleasure trying to contribute a bit.

**Comment by phweyland** [comment]
DT_SIGNAL_IMAGE_INFO_CHANGED used to update thumbnail group.

Doubt: in thumbtable.c there is no dt_control_signal_disconnect(). Is that OK ?

**Comment by AlicVB** [comment]
Yes, I think it's way better like that (more generic). I'll test in the evening. Thanks !

> Doubt: in thumbtable.c there is no dt_control_signal_disconnect(). Is that OK ?

Yes, there's only one instance of dt_thumbtable_t and it's alive until dt close

**Comment by TurboGit** [comment]
Clicking on the G icon does nothing with this patch. Is that me ?

**Comment by phweyland** [comment]
> Clicking on the G icon does nothing with this patch. Is that me ?

I assume you talk about image G icon.
Normally the PR doesn't impact the buttons themselves.
For my understanding, what does mean "with this patch" ? Do you patch your current branch with this commit ?

**Comment by TurboGit** [comment]
No I'm speaking about the top-right G icon to group or ungroup.

**Comment by TurboGit** [comment]
> For my understanding, what does mean "with this patch" ? Do you patch your current branch with this commit ?

No sure to see the question. Yes I'm on master, I locally merge this PR (patch) on master to test it before allowing it to be merged on GitHub.

**Comment by phweyland** [comment]
> No I'm speaking about the top-right G icon to group or ungroup.

... As long there is no group on lt it does nothing. But for me it expand/collapse the visible groups.

**Comment by phweyland** [comment]
> Yes I'm on master, I locally merge this PR (patch) on master to test it before allowing it to be merged on GitHub.

OK, to be in the same condition I need to rebase.

**Comment by phweyland** [comment]
Rebased. I confirm that on my side, the lt G button collapses/expands the groups on the view.

**Comment by TurboGit** [comment]
I have lot of strange behaviors recently :(

**Comment by TurboGit** [comment]
@AlicVB , @Nilvus : can you confirm that the G selector top-right of lighttable is working fine ? I cannot validate this PR on my side as the G is not working (doing nothing on my side). If you both confirm it means that there is something really wrong on my side that I need to understand. Otherwise we may be in a non-initialized variable kind of issue...

**Comment by phweyland** [comment]
```
{ "dt-image-info-changed", NULL, NULL, G_TYPE_NONE, g_cclosure_marshal_generic, 1, pointer_arg,
G_CALLBACK(_image_info_changed_destroy_callback), FALSE }, // DT_SIGNAL_IMAGE_INFO_CHANGED
```

In signal.c, line 102, I use g_cclosure_marshal_generic, but in some settings I can see different parameters. Should I use g_cclosure_marshal_VOID__POINTER instead ?

If yes, is line 94 still correct ?

**Comment by TurboGit** [comment]
> In signal.c, line 102, I use g_cclosure_marshal_generic, but in some settings I can see different parameters. Should I use g_cclosure_marshal_VOID__POINTER instead ?

No, this is correct. `g_cclosure_marshal_VOID__POINTER` does not exists in this file btw. I'm not sure it exists, but current code is ok and yours seems ok too.

**Comment by Nilvus** [comment]
> @AlicVB , @Nilvus : can you confirm that the G selector top-right of lighttable is working fine ? I cannot validate this PR on my side as the G is not working (doing nothing on my side). If you both confirm it means that there is something really wrong on my side that I need to understand. Otherwise we may be in a non-initialized variable kind of issue...

It works on my side (as it does), on master and with that PR rebased on master. Both icons (the one beside the star and help ones and the one on thumbnail. It seems you have a problem on your side.

**Comment by TurboGit** [comment]
That's very worrisome, it means that we have a non initialized value or memory corruption somewhere. I have checked that the G icon does run properly the callback on my side. So down the path there is something wrong but I cannot pinpoint it at the moment.

Given feedback I'll merge this has it looks ok to me and my issue is in anyway on master and with this PR.

**Comment by TurboGit** [comment]
Ok, issue found and now fixed. It was dependent on a wrong value in plugins/collection/filter_flags setting!

**Pull request merged**